### PR TITLE
SECURITY: allow UI endpoints to use access token authentication

### PR DESCRIPTION
**Pull request by slimsag**
Please see individual commits for details.

This change is for supporting https://github.com/sourcegraph/sourcegraph/pull/800

> This PR does not need to update the CHANGELOG because very minor changes, not worth mentioning.

### CODE REVIEW

File: cmd/frontend/internal/httpapi/auth.go [hunk 1]
```diff
@@ -18,8 +18,22 @@ func AccessTokenAuthMiddleware(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		w.Header().Add("Vary", "Authorization")
 
-		headerValue := r.Header.Get("Authorization")
+		// Secondary to the Authorization header (see below), access tokens may
+		// be specified via a query parameter (?token=<token>) OR via basic
+		// auth username (https://<token>@sourcegraph.com/foobar).
 		tokenParams, hasTokenParam := r.URL.Query()["token"]
+		tokenParam := ""
```
Comment (felixfbecker) [inline cmd/frontend/internal/httpapi/auth.go]: Maybe rename this to `urlToken` or `tokenFromURL` since it is not just from the query parameters anymore

**Comment by slimsag** [comment]
We decided to go with two separate routes, one being under `/.api` for token authentication and the other being under the UI route for cookie authentication (browser downloads).

**Comment by slimsag** [comment]
@nicksnyder looks like you have some pretty detailed suggestions for simplifying this code, I think we should always be looking to simplify code (_especially auth code_); why not send these improvements up as PRs? 😃

**Comment by nicksnyder** [comment]
https://github.com/sourcegraph/sourcegraph/pull/895

**Comment by beyang** [comment]
Passing token via URL param encourages bad behavior (since tokens are essentially passwords and you wouldn't embed your password in a URL param that might get stored in your browser history). Can we eliminate that?

The other question that comes to mind is, "Why don't GitHub or other code hosts enable token access to their non-api HTTP endpoints?" I think the answer is that token creation in GitHub requires a password (i.e., you can't use a token to create a new token). This is not the case for Sourcegraph, as we permit token creation through the API, but it does mean we'll need to be mindful if we ever do add some UI-based action that we want to require a password for. This change elevates tokens to be scoped passwords. I think that's fine, but we should make that clear––could you add something to the access tokens package doc (https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/access_tokens.go) that mentions this (that tokens should be treated as scoped passwords)?

**Pull request closed without merging**